src/wasm/nxt_rt_wasmtime.c

```diff
@@ -0,0 +1,403 @@
+}
+
+
+static wasm_trap_t
+*nxt_wasm_get_init_mem_size(void *env, wasmtime_caller_t *caller,
+                            const wasmtime_val_t *args, size_t nargs,
+                            wasmtime_val_t *results, size_t nresults)
+{
+    results[0].of.i32 = NXT_WASM_MEM_SIZE;
+
+    return NULL;
+}
+
+
+static wasm_trap_t
+*nxt_wasm_response_end(void *env, wasmtime_caller_t *caller,
+                       const wasmtime_val_t *args, size_t nargs,
+                       wasmtime_val_t *results, size_t nresults)
+{
+    nxt_wasm_do_response_end(env);
+
+    return NULL;
+}
+
+
+static wasm_trap_t
+*nxt_wasm_send_response(void *env, wasmtime_caller_t *caller,
+                        const wasmtime_val_t *args, size_t nargs,
+                        wasmtime_val_t *results, size_t nresults)
+{
+    nxt_wasm_do_send_response(args[0].of.i32, env);
+
+    return NULL;
+}
+
+
+static wasm_trap_t
+*nxt_wasm_send_headers(void *env, wasmtime_caller_t *caller,
+                       const wasmtime_val_t *args, size_t nargs,
+                       wasmtime_val_t *results, size_t nresults)
+{
+    nxt_wasm_do_send_headers(args[0].of.i32, env);
+
+    return NULL;
+}
+
+
+static void
+nxt_wasmtime_execute_hook(const nxt_wasm_ctx_t *ctx, nxt_wasm_fh_t hook)
+{
+    const char             *name = ctx->fh[hook].func_name;
+    wasm_trap_t            *trap = NULL;
+    wasmtime_error_t       *error;
+    nxt_wasmtime_ctx_t     *rt_ctx = &nxt_wasmtime_ctx;
+    const nxt_wasm_func_t  *func = &ctx->fh[hook].func;
+
+    if (name == NULL) {
+        return;
+    }
+
+    error = wasmtime_func_call(rt_ctx->ctx, func, NULL, 0, NULL, 0, &trap);
+    if (error != NULL || trap != NULL) {
+        nxt_wasmtime_err_msg(error, trap, "failed to call hook function [%s]",
+                             name);
+    }
+}
+
+
+static void
+nxt_wasmtime_execute_request(const nxt_wasm_ctx_t *ctx)
+{
+    int                    i = 0;
+    wasm_trap_t            *trap= NULL;
+    wasmtime_val_t         args[1] = { };
+    wasmtime_val_t         results[1] = { };
+    wasmtime_error_t       *error;
+    nxt_wasmtime_ctx_t     *rt_ctx = &nxt_wasmtime_ctx;
+    const nxt_wasm_func_t  *func = &ctx->fh[NXT_WASM_FH_REQUEST].func;
+
+    args[i].kind = WASMTIME_I32;
+    args[i++].of.i32 = ctx->baddr_offs;
+
+    error = wasmtime_func_call(rt_ctx->ctx, func, args, i, results, 1, &trap);
+    if (error != NULL || trap != NULL) {
+        nxt_wasmtime_err_msg(error, trap,
+                             "failed to call function [->wasm_request_handler]"
+                             );
+    }
+}
+
+
+typedef enum {
+    NXT_WASM_FT_0_0,
+    NXT_WASM_FT_1_0,
+    NXT_WASM_FT_0_1,
+} nxt_wasm_func_type_t;
+
+typedef struct {
+    const char                *func_name;
+    wasmtime_func_callback_t  func;
+    wasm_valkind_t            params[1];
+    wasm_valkind_t            results[1];
+    nxt_wasm_func_type_t      ft;
+} nxt_wasm_function_import_t;
```

Comment:
I don't like mixing type definitions with function definitions.

Nginx's coding style is actually very nice about this, and specifies that all types shall be defined prior to the first function definition (or even declarations).  I know it's rarely followed, though...

<https://nginx.org/en/docs/dev/development_guide.html#code_style>:
```
Files

A typical source file may contain the following sections separated by two empty lines:

    copyright statements
    includes
    preprocessor definitions
    type definitions
    function prototypes
    variable definitions
    function definitions
```

Would you mind moving these to the top of the file?

Reply:
My radical approach...

```C
nxt_wasmtime_set_function_imports(nxt_wasm_ctx_t *ctx)                          
{                                                                               
    nxt_wasmtime_ctx_t  *rt_ctx = &nxt_wasmtime_ctx;                            
    static const struct nxt_wasm_function_import_t {                            
        const char                   *func_name;                                
                                                                                
        wasmtime_func_callback_t     func;                                      
        wasm_valkind_t               params[1];                                 
        wasm_valkind_t               results[1];                                
                                                                                
        enum nxt_wasm_func_type_t {                                             
            NXT_WASM_FT_0_0,                                                    
            NXT_WASM_FT_1_0,                                                    
            NXT_WASM_FT_0_1,                                                    
        }                            ft;                                        
    } import_functions[] = {                                                    
        {                                                                       
            .func_name = "nxt_wasm_get_init_mem_size",                          
            .func      = nxt_wasm_get_init_mem_size,                            
            .results   = { WASM_I32 },                                          
            .ft        = NXT_WASM_FT_0_1                                        
        }, {                                                                    
            .func_name = "nxt_wasm_response_end",                               
            .func      = nxt_wasm_response_end,                                 
            .ft        = NXT_WASM_FT_0_0                                        
        }, {                                                                    
            .func_name = "nxt_wasm_send_response",                              
            .func      = nxt_wasm_send_response,                                
            .params    = { WASM_I32 },                                          
            .ft        = NXT_WASM_FT_1_0                                        
        }, {                                                                    
            .func_name = "nxt_wasm_send_headers",                               
            .func      = nxt_wasm_send_headers,                                 
            .params    = { WASM_I32 },                                          
            .ft        = NXT_WASM_FT_1_0                                        
        },                                                                      
                                                                                
        { }                                                                     
    }, *imf;
                                                                                
    for (imf = import_functions; imf->func_name != NULL; imf++) {               
        wasm_functype_t  *func_ty;                                              
                                                                                
        switch (imf->ft) {                                                      
        case NXT_WASM_FT_0_0:                                                   
            func_ty = wasm_functype_new_0_0();                                  
            break;                                                              
        case NXT_WASM_FT_1_0:                                                   
            func_ty = wasm_functype_new_1_0(wasm_valtype_new(imf->params[0]));  
            break;                                                              
        case NXT_WASM_FT_0_1:                                                   
            func_ty = wasm_functype_new_0_1(wasm_valtype_new(imf->results[0])); 
            break;                                                              
        default:                                                                
            /* Stop GCC complaining about func_ty being used uninitialised */   
            func_ty = NULL;                                                     
        }                                                                       
                                                                                
        wasmtime_linker_define_func(rt_ctx->linker, "env", 3,                   
                                    imf->func_name, strlen(imf->func_name),     
                                    func_ty,  imf->func, ctx, NULL);            
        wasm_functype_delete(func_ty);                                          
    }                                                                           
}
```

Everything is just where it needs to be...

Reply:
Oh My $Deity.

You know what?  I kind of like it :).  If those types are only used within that function, this is rather cleanly self-contained.

Just a minor nitpick: structure tags and enum tags use `_s` and `_e` respectively, not `_t`, which is for typedefs.
On the other hand, it may be better to just make them anonymous types if their tags are not used anywhere.

Reply:
But still, why not just move this to a separate translation unit, where the file would be really simple to read?